ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/MessageListenerProxy.java

```diff
@@ -0,0 +1,74 @@
+package org.jboss.narayana.jta.jms;
+
+import com.arjuna.ats.jta.logging.jtaLogger;
+
+import javax.jms.Message;
+import javax.jms.MessageListener;
+import javax.jms.XASession;
+import javax.transaction.SystemException;
+import javax.transaction.Transaction;
+import javax.transaction.TransactionManager;
+import javax.transaction.Transactional;
+import java.lang.reflect.Method;
+
+public class MessageListenerProxy implements MessageListener {
+
+    private MessageListener messageListener;
+    private TransactionManager transactionManager;
+    private XASession xaSession;
+
+    static boolean isDeclaredTransactional(MessageListener messageListener) {
+        try {
+            if (messageListener == null) {
+                return false;
+            }
+
+            Class<? extends MessageListener> listenerClass = messageListener.getClass();
+            if (listenerClass.isAnnotationPresent(Transactional.class)) {
+                return true;
+            } else {
+                Method onMessageMethod = listenerClass.getDeclaredMethod("onMessage", Message.class);
+                return onMessageMethod.isAnnotationPresent(Transactional.class);
+            }
+        } catch (NoSuchMethodException e) {
+            // Should not occur, as messageListener is an instance of javax.jms.MessageListener
+        }
+        return false;
+    }
+
+    public MessageListenerProxy(MessageListener messageListener, XASession session) {
+        this(messageListener, session, com.arjuna.ats.jta.TransactionManager.transactionManager());
+    }
+
+    /**
+     * Mainly for tests
+     */
+    MessageListenerProxy(MessageListener messageListener, XASession session, TransactionManager transactionManager ) {
+        this.messageListener = messageListener;
+        this.transactionManager = transactionManager;
+        this.xaSession = session;
+    }
+
+    @Override
+    public void onMessage(Message jmsMessage) {
```

Comment:
Thanks, if its not spec covered we shouldn't go this route. We could investigate moving to a thread pool that does synchronous receives but calls back to message listeners. It's likely not as efficient as the broker doing that for us though.

Another option would be to look at how to integrate with the resource adapter outside of a container. The links I have provided before show the code we would need (beforeDelivery/afterDelivery). Possibly we would need to try to include IronJacamar standalone to get that working.

You can call createSession within a dummy transaction to go through the route you needed in ConnectionProxy. I don't think you should rely on the MessageListenerProxy behaviour if you aren't covered by spec as although it may work with todays message brokers they are free to change in the future and you would likely not spot the change in behaviour until it was too late and you had data integrity issues.